src/codegen.c

```diff
@@ -447,16 +453,30 @@ static void codegen_expr(CodegenContext *ctx, Node *expr) {
 
   /// Literal expression. Only integer literals are supported for now.
   case NODE_LITERAL:
-    if (expr->literal.type != TK_NUMBER) DIAG(DIAG_SORRY, expr->source_location, "Emitting non-integer literals not supported");
+    if (expr->literal.type == TK_NUMBER) expr->ir = ir_immediate(ctx, expr->type, expr->literal.integer);
+    else if (expr->literal.type == TK_STRING) {
+      char buf[48] = {0};
+      static size_t string_literal_count = 0;
+      int len = snprintf(buf, 48, "__str_lit%zu", string_literal_count++);
+      // TODO: Two options. Since we can't currently assign to static
+      // variables with compile-time known constants, it means that
+      // strings go uninitialised... So what would really be ideal
+      // is if we could just, ya know, do that. (i.e. `.string` as
+      // directive or asciz or whathaveyou). Our other option is to
+      // emit (imm+store) pairs for every byte...
+      expr->ir = ir_create_static(ctx, expr->type, as_span(string_create(buf)));
+    }
+    else DIAG(DIAG_SORRY, expr->source_location, "Emitting non-integer literals not supported");
     // TODO: SEMA should probably have already lowered integer_literal type, so we *should* have a type already available on the literal node...
-    expr->ir = ir_immediate(ctx, expr->type, expr->literal.integer);
     return;
 
   /// Variable reference.
   case NODE_VARIABLE_REFERENCE:
     expr->ir = ir_load(ctx, expr->var->val.node->ir);
     // TODO: Be smarter about when an array should decay to a pointer or not.
     //       Maybe it never should, and this should be implemented per backend?
+    // "I’d just emit a load of the array and have the backend
+    // deal w/ copying 1000 ints." ~ Sirraide
     if (expr->ir->type->kind == TYPE_ARRAY)
```

Comment:
To expound on this: If we get here, we should assume that we have a case like e.g. `a := b`  where `b` is an array type, and we’re trying to emit `b`, in which case we just emit a load and be done w/ it. 

If we’re dealing w/ something like `b[0]` instead (or `a` in `a := b`, or `&b`), then we should check for that case when codegenning the subscript operator (or the assignment or addressof operator) and not even try and emit `b` in the first place iff it is an array. This is also how cases like `@foo := bar` are handled because dereferencing suffers from all the same problems...

Reply:
> not even try and emit b in the first place iff it is an array

This works fine in simple cases (e.g. `&b`), `b` has to be an lvalue anyway for us to be able to take its address, and we don’t need to emit a load to it to know its address.

Where this probably doesn’t work atm is if we have something like `(if cond { a } else { b }) := c`, but then again, I’m not sure this case would work currently, even if `a` and `b` are regular variables... so perhaps we might have to change it to where it only actually loads the value once we know that is used by some other instruction (and used AS A VALUE, not as an address to e.g. store to). I’m not sure though if this should happen by somehow not generating a load when we only want the lvalue or by removing the load afterwards if it’s clear that it’s unused. 

Optimisations would take care of the latter anyway, so perhaps a solution would be to make optimising dead loads mandatory? Only trivially dead loads, i.e. loads that have no users at all, that is. We wouldn’t have to account for loads that are indirectly dead because, although they are used to compute some value, *that* value is also dead code. That doesn’t matter in this case since if we were to emit a load when we actually just need an lvalue, then the load proper would be completely unused.

Reply:
The top-down approach (i.e. checking if a value needs to be loaded or if we just want its address) would entail emitting potentially arbitrarily complex and nested expressions while not actually loading anything and instead just collecting the lvalues. All of this is just an idea tho; I haven’t thought too much about this in a while and it also feels like I’m overcomplicating things here candidly...

Reply:
This entire thing is basically more or less related to getting something like this to work properly:

![image](https://user-images.githubusercontent.com/74590115/216425404-70970ee2-0fa5-4357-9144-1e5ecd5f0401.png)

I guess there is a finite number of expressions that may return an lvalue (e.g. `if` expressions may, but `while` loops can’t), so one option would be to leave `codegen_expr` untouched and add a separate `codegen_lvalue` or sth that gets called e.g. when emitting an assignment or array subscript, and instead of loading the value, it just ‘loads’ the address. For example, the value of the if expression in `(if a { @b ⟩ else { @c }) := 5` at runtime would be the *address* of `b` or `c`, to which we could then emit an (indirect) store. This entire thing is basically just about handling lvalues properly. Once we can codegen e.g. that if expression, then arrays should just work because they’re basically the same thing...

Reply:
> how do I actually emit a load of a pointer (i.e. the pointer) without dereferencing it?

Do I need to actually emit a load of a pointer to the pointer?? I feel like that's a bit scuffed

Reply:
Right, you’re going about this wrong again. You don’t load the address of an lvalue. That’s why I said ‘load’, not load. The address of *any* lvalue is known at compile time. To codegen e.g. `(if a { @b ⟩ else { @c }) := 5`, assuming all variables are local variables, we need to emit this:
```
    ;; test `a` etc.
    jnz b_block
    jmp c_block
b_block:
    lea rdi, [rbp + N] ;; where `N` is the stack offset of `b`
    jmp end
c_block:
    lea rdi, [rbp + M] ;; where `M` is the stack offset of `c`
end:
    mov [rdi], 5
```

Reply:
The question is, how do we best represent this in IR and how do we even codegen the IR in the first place

Reply:
`IR_ADDRESSOF`?

Reply:
So I guess the IR we need to emit for this would be:
```
    %b = alloca 8
    %c = alloca 8
    ;; ...
    br.cond %a, bb_b, bb_c
bb_b:
    br bb_end
bb_c:
    br bb_end
bb_end:
    %if = phi [bb_b: %b, bb_c: %c]
    store into %if, 5
```

Reply:
> `IR_ADDRESSOF`?

See the problem is that `addressof` as a concept doesn’t really make sense because you can only ‘take’ the address of something if it’s an lvalue, and if it is an lvalue, then you already know the address because an lvalue refers to an address in memory

Reply:
We already have special handling for some cases:
https://github.com/LensPlaysGames/FUNCompiler/blob/974bdff6f8a86e336a4c723f70c67a59867fa206/src/codegen.c#L342-L358

Reply:
We’d have to do something similar when codegenning subscripts if the LHS is an array (what’s crucial is that we don’t try and codegen the LHS first).

Reply:
For the if expression above, we could e.g. add an `as_lvalue` parameter or whatever to `codegen_expr()` which, if true, would mean that
    1. Trying to codegen something that is not an lvalue would be an error (sema should have already caught this)
    2. When emitting a ‘load’, we instead return the address as a value.

So I suppose, you are right in that we need something like IR_ADDRESSOF, not so much because we’re ‘taking’ the address of something, but rather because we’re storing the address of something as a value.

Reply:
The special cases I linked above would then be handled wherever we’d normally emit loads (e.g. codegen for unary operator `@`, codegen for arrays etc.)

Reply:
As for the value of that `as_lvalue` parameter, we would call:
    - `codegen_expr(..., false)` if we’re codegenning the root or a function;
    - `codegen_expr(..., true)` if we’re codegenning e.g. the LHS of a store;
    - `codegen_expr(..., as_lvalue)` otherwise, passing along the whatever value it currently has.

Reply:
I wouldn’t call this the cleanest solution, but it is *a* solution I suppose...

Reply:
As for arrays in particular, we’d emit a load if `as_lvalue` is `false`, and ‘take’ the address if `as_lvalue` is `true`.

Reply:
I'm beginning to understand somewhat now; a major point of confusion was the phrasing "not emit a load" or similar (I'll be the first to admit that is on me). I was confused as to where the value would come from if it's never loaded. But it seems like what you are referencing is the fact that the "load" has *already* been emitted, and therefore it's not necessary to emit it again. Instead, we don't emit a load *because one already exists*; we can just use that instructions value. What helped me see this was you pointing out the current handling of dereference, as I know how that works and why it's needed.

Reply:
Can

```
    %b = alloca 8
    %c = alloca 8
    ;; ...
    br.cond %a, bb_b, bb_c
bb_b:
    br bb_end
bb_c:
    br bb_end
bb_end:
    %if = phi [bb_b: %b, bb_c: %c]
    store into %if, 5
```

be turned into

```
    br.cond %a, bb_b, bb_c
bb_b:
    %b = alloca 8
    br bb_end
bb_c:
    %c = alloca 8
    br bb_end
bb_end:
    %if = phi [bb_b: %b, bb_c: %c]
    store into %if, 5
```

Not saying it should, really just wondering if they are equivalent

Reply:
They’re not really the same. The latter is what you’d get if you did this: `if a { b : integer } else { c : integer }`.

Reply:
An alloca is a variable *declaration*

Reply:
The codegen for the `if` is everything after and including the `br.cond`. It’s assumed that `b` and `c` are defined earlier on in the function; I included the `alloca`s mostly to make clear that these are supposed to be variables